stdlib/dynarray.ml

```diff
@@ -0,0 +1,334 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                            Simon Cruanes                               *)
+(*                                                                        *)
+(*   Copyright 2022 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+type 'a t = {
+  mutable size : int;
+  mutable arr : 'a array;
+}
+
+(* TODO: move to runtime? bypass write barrier *)
+let[@inline] fill_ (a:_ array) i ~filler : unit =
+  Array.set a i filler
+
+(* TODO: move to runtime? bypass write barrier *)
+let[@inline] fill_with_junk_ (a:_ array) i len ~filler : unit =
+  Array.fill a i len filler
+
+let create () = {
+  size = 0;
+  arr = [| |];
+}
+
+let make n x = {
+  size=n;
+  arr=Array.make n x;
+}
+
+let init n f = {
+  size=n;
+  arr=Array.init n f;
+}
+
+let blit v1 i1 v2 i2 len =
+  if i1<0 || i2<0 || i1+len >= v1.size || i2 + len >= v2.size then
+    invalid_arg "Dynarray.blit";
+  Array.blit v1.arr i1 v2.arr i2 len
+
+(* is the underlying array empty? *)
+let[@inline] array_is_empty_ v =
+  Array.length v.arr = 0
+
+(* next capacity, if current one is [n]. Roughly use [n * 1.5], because it
+   provides the good behavior of amortized O(1) number of allocations
+   without wasting too much memory in the worst case. *)
+let[@inline] next_grow_ n =
+  min Sys.max_array_length (1 + n + n lsr 1)
```

Comment:
If the user has called `fit_capacity` on a very small array, some degenerate behavior will happen next, since this function will return a grown size smaller than the minimal size (i.e., 4). This can be avoided by prepending `max 4`, or by changing the formula a bit, e.g., `n + (n + 5) lsr 1`.

By the way, it is folklore that, with a non-compacting garbage collector, the optimal growth ratio is the golden ratio. So, choosing 1.5 in practice is less arbitrary than it looks like.